### PR TITLE
[Master] Decouple Master and Master extensions

### PR DESCRIPTION
**Pull request by ddmbr**
This is to continue the work on the old repo. In particular, this patch
decouples the extensions (e.g., MongoDB module) from Master so that
these modules can extend Master in a non-intrusive way.

### CODE REVIEW

**Comment by ddmbr** [comment]
@Yuzhen11 @kygx-legend Please review

File: master/mongodb_assigner.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
+#ifdef WITH_MONGODB
```
Comment (kygx-legend) [inline master/mongodb_assigner.cpp]: Could you add copyright information?

File: master/master.cpp [hunk 2]
```diff
@@ -63,6 +61,10 @@ Master::Master() {
     lb_step = -2;
     partition_to_donate.resize(partition_table.size(), false);
     debug_num_broadcast = 0;
+
+    for(auto setup_handler : external_setup_handlers) {
```
Comment (kygx-legend) [inline master/master.cpp]: Lack of a space after `for`.

File: master/master.hpp [hunk 3]
```diff
@@ -26,29 +26,36 @@
...
     inline zmq::socket_t* get_socket() const { return resp_socket; }
     inline const std::string& get_cur_client() const { return cur_client; }
 
-    inline void register_handler(uint32_t msg_type, std::function<void()> handler) {
-        external_handlers[msg_type] = handler;
+    inline void register_main_handler(uint32_t msg_type, std::function<void()> handler) {
+        external_main_handlers[msg_type] = handler;
+    }
+
+    inline void register_setup_handler(std::function<void()> handler) {
```
Comment (kygx-legend) [inline master/master.hpp]: Could you run `scripts/clang-format.sh` and then commit again? There're some differences output in [travis ci](https://travis-ci.org/husky-team/husky/builds/161270459)

File: master/hdfs_assigner.cpp [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    base::log_msg(host+" => "+ret.first+"@"+std::to_string(ret.second));
+}
+
+void HDFSBlockAssigner::master_setup_handler() {
+    init_hdfs(Context::get_param("hdfs_namenode"), Context::get_param("hdfs_namenode_port"));
+    num_workers_alive = Context::get_worker_info()->get_num_workers();
+}
+
+HDFSBlockAssigner::~HDFSBlockAssigner() {}
```
Comment (Yuzhen11) [inline master/hdfs_assigner.cpp]: May look nicer if just use = default in the hpp file.

File: master/master.cpp (hunk 2 above)
Comment (Yuzhen11) [inline master/master.cpp]: auto& may be better

Comment (ddmbr) [inline master/master.cpp]: @Yuzhen11 These are std::function, so using reference won't make it more efficient? We can verify it anyway.

**Comment by ddmbr** [comment]
Please check again @Yuzhen11 @kygx-legend

**Comment by Yuzhen11** [comment]
LGTM

File: master/mongodb_assigner.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    if (!split.is_valid())
+        return;
+
+    end_count_++;
+    splits_end_.push_back(split);
+}
+
+static MongoSplitAssigner mongo_split_assigner;
+
+}
```
Comment (kygx-legend) [inline master/mongodb_assigner.cpp]: `}  // namespace husky`

File: master/mongodb_assigner.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+void MongoSplitAssigner::recieve_end(MongoDBSplit& split) {
+    if (!split.is_valid())
+        return;
+
+    end_count_++;
+    splits_end_.push_back(split);
+}
+
+static MongoSplitAssigner mongo_split_assigner;
```
Comment (kygx-legend) [inline master/mongodb_assigner.cpp]: Could you put line 160 in line 30? The static variable feels like to be hidden in the end of this file.

File: master/hdfs_assigner.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+/// Return the number of workers who have finished reading the files in
+/// the given url
+int HDFSBlockAssigner::get_num_finished(std::string& url) { return finish_dict[url]; }
+
+/// Use this when all workers finish reading the files in url
+void HDFSBlockAssigner::finish_url(std::string& url) { files_locality_dict.erase(url); }
+
+static HDFSBlockAssigner hdfs_block_assigner;
```
Comment (kygx-legend) [inline master/hdfs_assigner.cpp]: Put the static variable before functions.

**Pull request merged**